client/src/reducers/index.js

```diff
@@ -1,21 +1,33 @@
-import { 
-  SET_USER, 
-} from "../constants/action-types";
+import { AUTH_LOGIN, AUTH_LOGOUT, SET_USER } from "../constants/action-types";
 
 const initialState = {
+  accessToken: null,
+  isAuthenticated: false,
   user: null,
 };
 
 function rootReducer(state = initialState, action) {
   switch (action.type) {
+    case AUTH_LOGIN:
+      return {
+        ...state,
+        accessToken: action.payload.token,
+        isAuthenticated: true,
+      };
+    case AUTH_LOGOUT:
+      return {
+        ...state,
+        accessToken: null,
+        isAuthenticated: false,
+      };
     case SET_USER:
```

Comment:
Why is `SET_USER` a separate action, and not a part of login and logout? Not sure when we would use this separate of auth related actions.

Reply:
yeah that's the thing... it's maybe not really standard, we were only using "SET_USER" before, someone added it, maybe @markthien who added redux and thought this was usually what happens when you login, you get the user info
But in this case, we have a call to `/api/authenticate` (which is currently used for both login & signup but I would like to split them up on the backend too), which returns a token from Auth0
Auth0 can also retrieve the full user object back, but after discussing and looking at some documentation, it seems better to have our own User model in mongodb and store most data there, so the idea is that Auth0 will do just the authentication part (as I mentioned in description of the PR) 
This also means that we don't have a "user" object when we call `authenticate` but just a token. 

We could also add that later on the server side, if we get success response from Auth0 that we also check our mongo and return a user, but maybe it's better to make a separate request to `/api/ucers/current` or similar for that... and then it will need separate redux action and state update... this is why i opened it as a draft PR because it's not really set in stone yet.

Reply:
Yea I'm alright about separating the between using our own User model to store user data and having auth0 just returning auth tokens. It just seems like extra work to not use auth0's full user object. I probably missed the discussion but why aren't we using auth0's user object?

I thought we had a route that gets the current user? Either way I agree it would be handy to have one. I guess that's where we would dispatch  `SET_USER` to  `api/users/current` to set the current user in the frontend.

Reply:
About using our own User model with Auth0 just as authentication, the main thing would be the integration between the user and all the other data. With a User model of our own at Mongo, we can easily query it alongside other data in a single request (such as bring the author's name and location along with the post query for the feed) or even use the user's info in more complex aggregation pipelines. With the user only inside Auth0, we would need a separate request for that and it would need application code for integrating the info with the data that is inside Mongo.

When I was doing some research on how people usually approach this and found [this discussion](https://community.auth0.com/t/how-to-store-user-information/22148) at the Auth0 forum that seems to be similar to our case (trying to build twitter clone and wanted to know the best approach for storing User data). At the post they present 2 approaches, both keeping a User model at our side, one kept both models in sync and the other used Auth0 only for auth. Long story short, we found that keeping only the authentication at Auth0 side may be the best case to ease the backend pain of keeping both models in sync.

Reply:
> I thought we had a route that gets the current user? Either way I agree it would be handy to have one. I guess that's where we would dispatch `SET_USER` to `api/users/current` to set the current user in the frontend.

Yeah there was already a route made initially but we probably have to clean that up a bit. And yes like @Naraujo13 mentioned, if we do queries for posts and organizations etc. we might need to fetch some user info regularly and then it seems annoying if we have to call the Auth0 API every time, and it can make things like aggregations for mongo queries also impossible in this way, we could also cache user info in mongo but that's adding also going to add more complexity.

So yea if we have separate /login and /register routes, and use `/api/users` for all the CRUD operations for users... only disadvantage is having to make 2 request sometimes e.g. `/auth/login` gets token and then `/iusers/current` gets user info (if we used GraphQL we could do something like getting user at login easier without breaking the REST "rules"). We could also consider to add extra logic in the backend that we check if user exists in our db on /login or /signup.

So if you login, and Auth0 returns that email is not verified => verify email first
If you login and we can't find your user in our db => create profile first

Just need to check now if it's all going to work like this and if we forgot about something.

Reply:
so one caveat right now is that calling `/signup` and then `authenticate` gets an OAuth token but not the user info from Auth0 so we have to add an additional `getUserInfo` call to make sure email is confirmed after login / signup... so that's a drawback of not using same User db, and i have to check some more of the security issues but that's out of scope here